### DISCUSSION TITLE
__class__ and type()

### DISCUSSION THREAD

**Discussion by mattgreen** [PyO3/pyo3#2189] [Questions] [answered]
I'm working on a Python binding to a DSL. It exposes proxies to "objects" in Rust code, which are structures with an `ObjectKind` field denoting the type (Array, Int, etc). They are exposed to Python as subclasses of an `Object` class, so Python code sees an `Int` with a value property that queries the underlying Rust value.

The Rust-side objects can change their type during runtime. This causes their Python proxies to get out of sync: an `Int` class in Python could be referencing an `ObjectKind::Array`!

After too much head-scratching, I realized PyO3 lets you override `__class__` as a getter, at which point I could return the class that the object should be. Pretty cool! However, I noticed that `type(obj)` still seems to return the class it was originally created with, instead of querying my implementation of `__class__`.

Is this a PyO3 thing? It smells like the Python interpreter querying private state on the object vs asking the object what type it is. It isn't a dealbreaker but I figured I'd ask.

**Comment by davidhewitt** [comment, marked as answer]
> Is this a PyO3 thing? It smells like the Python interpreter querying private state on the object vs asking the object what type it is. It isn't a dealbreaker but I figured I'd ask.

Python interpreter querying private state on the object is exactly what it is. Here's the relevant line of C in the implementation of the `type` builtin:

https://github.com/python/cpython/blob/87af12bff33b3e7546fa26158b7d8680ecb6ecec/Objects/typeobject.c#L3343

**Reply by mattgreen** [reply]
Thank you! I see now, I need to also `setattr(obj, '__class__', new)` when I observe the type change.
